### PR TITLE
Feature/test driven development

### PR DESCRIPTION
**Pull request by skyv26**
# Milestone #1 - Test Driven Development

In this feature branch, @skyv26 and @WalterOkumu both wrote the test in order to develop. All the things that we have implemented are as follows:

- Created a class called `Solver`.
- Created a method called factorial that takes one argument, an integer `N`, and returns the [factorial](https://en.wikipedia.org/wiki/Factorial) for that number. The factorial is the multiplication of all integers from 1 to `N` and has the special case that the factorial of 0 is 1. This method only accepts 0 and positive integers, so if a negative integer is given it should raise an exception.
- Created a method called reverse that takes one argument, a string word, and returns the word reversed (e.g. if the word is `hello` it returns `olleh`).
- Created a method called `fizzbuzz` that takes one argument, an integer `N`, and returns a string. The returned string is constructed following these rules:
  - When N is divisible by 3, return `fizz`.
  - When N is divisible by 5, return `buzz`.
  - When N is divisible by 3 and 5, return `fizzbuzz`.
  - Any other case, return N as a string (e.g. say N is 7 then return `7`).
- Resolved the `linters` error.
- Used good practices and design patterns.
- Updated the `README.md` file.

### CODE REVIEW

File: spec/solver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      it 'will test the `factorial` of 3 should be 6' do
+        expect(@solver.factorial(3)).to eql(6)
+      end
+
+      it 'will test the `factorial` of 0 should be 0' do
+        expect(@solver.factorial(0)).to eql(0)
+      end
+      #   it 'will test the `factorial` of negative number should raise error' do
+      #     expect {@solver.factorial(1)}.to raise_error(RuntimeError)
+      #   end
```
Comment (ShahierNashaat) [inline spec/solver_spec.rb]: - [x] Kindly notice that you should add those two test cases, the Negative value which raises an exception as an expectation, and the Special case - when input is 0, the result should equal 1.

```ruby
it 'will test the `factorial` of 0 should be 0' do
    expect(@solver.factorial(0)).to eql(1)
end
it 'will test the `factorial` of negative number should raise error' do
     expect {@solver.factorial(-5)}.to raise_error(RuntimeError)
end
```

File: solver.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  def initialize
+    @exception = ExceptionHandler.new
+  end
+
+  def factorial(num)
+    @exception.msg('factorial') if num.negative?
+    return num if [1, 0].include?(num)
+
+    num * factorial(num - 1)
+  end
```
Comment (ShahierNashaat) [inline solver.rb]: - [x] Kindly notice that the factorial of 0 is 1 so it should return 1, not the num.

```ruby
def factorial(num)
    @exception.msg('factorial') if num.negative?
    return 1 if [1, 0].include?(num)

    num * factorial(num - 1)
end
```

**Review by ShahierNashaat — CHANGES_REQUESTED** [review]
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 💯 

- All tests are passing ✔️
- Correct use of GitFlow ✔️
- Descriptive commit messages ✔️
- Professional looking pull request ✔️
- Your readme looks professional ✔️
- Your code is clean and well documented ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/solver_spec.rb (hunk 1 above)
Comment (skyv26) [reply spec/solver_spec.rb]: @ShahierNashaat Thanks. solved

File: solver.rb (hunk 2 above)
Comment (skyv26) [reply solver.rb]: @ShahierNashaat Thanks. Solved...

**Review by t-yanick — APPROVED** [review]
Hi @skyv26, 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good commit messages
:heavy_check_mark: Descriptive Readme file

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

### N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
